dali/util/dynlink_cuda.h

```diff
@@ -0,0 +1,1972 @@
+typedef CUresult  tcuMemcpyDtoH(void *dstHost, CUdeviceptr srcDevice, unsigned int ByteCount);
+
+// device <-> device memory
+typedef CUresult  tcuMemcpyDtoD(CUdeviceptr dstDevice, CUdeviceptr srcDevice, unsigned int ByteCount);
+
+// device <-> array memory
+typedef CUresult  tcuMemcpyDtoA(CUarray dstArray, unsigned int dstOffset, CUdeviceptr srcDevice, unsigned int ByteCount);
+typedef CUresult  tcuMemcpyAtoD(CUdeviceptr dstDevice, CUarray srcArray, unsigned int srcOffset, unsigned int ByteCount);
+
+// system <-> array memory
+typedef CUresult  tcuMemcpyHtoA(CUarray dstArray, unsigned int dstOffset, const void *srcHost, unsigned int ByteCount);
+typedef CUresult  tcuMemcpyAtoH(void *dstHost, CUarray srcArray, unsigned int srcOffset, unsigned int ByteCount);
+
+// array <-> array memory
+typedef CUresult  tcuMemcpyAtoA(CUarray dstArray, unsigned int dstOffset, CUarray srcArray, unsigned int srcOffset, unsigned int ByteCount);
+#endif
+
+// 2D memcpy
+
+typedef CUresult  tcuMemcpy2D(const CUDA_MEMCPY2D *pCopy);
+typedef CUresult  tcuMemcpy2DUnaligned(const CUDA_MEMCPY2D *pCopy);
+
+// 3D memcpy
+
+typedef CUresult  tcuMemcpy3D(const CUDA_MEMCPY3D *pCopy);
+
+/************************************
+ **
+ **    Asynchronous Memcpy
+ **
+ ** Any host memory involved must be DMA'able (e.g., allocated with cuMemAllocHost).
+ ** memcpy's done with these functions execute in parallel with the CPU and, if
+ ** the hardware is available, may execute in parallel with the GPU.
+ ** Asynchronous memcpy must be accompanied by appropriate stream synchronization.
+ **
+ ***********************************/
+
+// 1D functions
+#if __CUDA_API_VERSION >= 3020
+// system <-> device memory
+typedef CUresult  tcuMemcpyHtoDAsync(CUdeviceptr dstDevice,
+                                             const void *srcHost, size_t ByteCount, CUstream hStream);
+typedef CUresult  tcuMemcpyDtoHAsync(void *dstHost,
+                                             CUdeviceptr srcDevice, size_t ByteCount, CUstream hStream);
+
+// device <-> device memory
+typedef CUresult  tcuMemcpyDtoDAsync(CUdeviceptr dstDevice,
+                                             CUdeviceptr srcDevice, size_t ByteCount, CUstream hStream);
+
+// system <-> array memory
+typedef CUresult  tcuMemcpyHtoAAsync(CUarray dstArray, size_t dstOffset,
+                                             const void *srcHost, size_t ByteCount, CUstream hStream);
+typedef CUresult  tcuMemcpyAtoHAsync(void *dstHost, CUarray srcArray, size_t srcOffset,
+                                             size_t ByteCount, CUstream hStream);
+
+#else
+// system <-> device memory
+typedef CUresult  tcuMemcpyHtoDAsync(CUdeviceptr dstDevice,
+                                             const void *srcHost, unsigned int ByteCount, CUstream hStream);
+typedef CUresult  tcuMemcpyDtoHAsync(void *dstHost,
+                                             CUdeviceptr srcDevice, unsigned int ByteCo
```

Comment:
Luckily, that's not much of a problem because we end up always calling the initializers on the main thread first. But that's true, we should have thread-safety mechanism anyway, will do

Reply:
Done